### PR TITLE
[apex] Add a rule to prevent use of non-existent annotations

### PR DESCRIPTION
**Pull request by anand13s**
This rule seeks to warn against using non-existent annotations in apex source code. A 'non existent' annotation is one not defined in the [apex language documentation annotations section. ](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation.htm)

This is done to prevent apex source code from being broken by newer Apex annotations in the Salesforce language that possess the same signature as a non existent one.

### CODE REVIEW

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/AbstractApexRule.java [hunk 1]
```diff
@@ -96,16 +96,21 @@
 import net.sourceforge.pmd.lang.ast.Node;
 import net.sourceforge.pmd.lang.rule.AbstractRule;
 import net.sourceforge.pmd.lang.rule.ImmutableLanguage;
+import net.sourceforge.pmd.lang.rule.properties.IntegerProperty;
 import net.sourceforge.pmd.renderers.CodeClimateRule;
 
 public abstract class AbstractApexRule extends AbstractRule
         implements ApexParserVisitor, ImmutableLanguage, CodeClimateRule {
 
+    public static final IntegerProperty CODACY_MINUTES_TO_FIX = new IntegerProperty("codacy_minutes_to_fix",
```
Comment (jsotuyod) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/AbstractApexRule.java]: the codacy properties are unrelated to the object of this PR, please revert this changeset

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/style/AvoidNonExistentAnnotationsRule.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
+package net.sourceforge.pmd.lang.apex.rule.style;
```
Comment (jsotuyod) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/style/AvoidNonExistentAnnotationsRule.java]: the rule seems like a better fit for the `errorprone` category (code that is likely to break) rather than a code style preference. Consider moving it there.

File: pmd-apex/src/main/resources/rulesets/apex/ruleset.xml [hunk 3]
```diff
@@ -1,128 +1,180 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <ruleset xmlns="http://pmd.sourceforge.net/ruleset/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="Default ruleset used by the CodeClimate Engine for Salesforce.com Apex" xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
+
```
Comment (jsotuyod) [inline pmd-apex/src/main/resources/rulesets/apex/ruleset.xml]: changes to this file are unrelated to the object of this PR, please revert

File: pmd-apex/src/main/resources/rulesets/apex/style.xml [hunk 4]
```diff
@@ -132,4 +132,24 @@ Many interfaces (e.g. Batch) required global modifiers in the past but don't req
     </example>
   </rule>
 
+<rule name="AvoidNonExistentAnnotations"
```
Comment (jsotuyod) [inline pmd-apex/src/main/resources/rulesets/apex/style.xml]: rulesets are deprecated. New rules should be added to categories. In particular, this looks like a fit for the [`errorprone` category](https://github.com/pmd/pmd/blob/master/pmd-apex/src/main/resources/category/apex/errorprone.xml)

File: pmd-apex/src/main/resources/rulesets/apex/style.xml [hunk 5]
```diff
@@ -132,4 +132,24 @@ Many interfaces (e.g. Batch) required global modifiers in the past but don't req
...
+          since="5.5.5"
```
Comment (jsotuyod) [inline pmd-apex/src/main/resources/rulesets/apex/style.xml]: the next release to include new rules will be 6.1.0

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/style/xml/AvoidNonExistentAnnotations.xml [hunk 6]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jsotuyod) [inline pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/style/xml/AvoidNonExistentAnnotations.xml]: consider adding test scenarios where valid annotations are not flagged

File: pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/InlinePragmaProcError.pls [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
-  begin
-    PRAGMA AUTONOMOUS_TRANSACTION;
-    bno74.log_hentglass_request(p_hentglass_request
-                               ,v_logging_req_seq_no);
-    COMMIT;
-   end do_transaction;
-
-end inline_praqma_error;
-/
+create or replace package inline_praqma_error is
```
Comment (jsotuyod) [inline pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/InlinePragmaProcError.pls]: this changeset is unrelated to the object of the PR, please revert

File: pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/IsOfType.pls [hunk 8]
```diff
@@ -1,23 +1,23 @@
...
-end if;
-
-IF ioFPOobj IS NOT OF TYPE (ONLY FPOGE_OBJ) THEN
-loFPOGE_OBJ:=treat(ioFPOobj AS FPOGE_OBJ);
-end if;
-
-loFPOGE_OBJ:=SELECT A FROM persons p WHERE IS OF TYPE (employee_t);
-loFPOGE_OBJ:=SELECT A FROM persons p WHERE IS NOT OF TYPE (ONLY employee_t, other_t);
-
+PROCEDURE IsOfType (
```
Comment (jsotuyod) [inline pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/IsOfType.pls]: this changeset is unrelated to the object of the PR, please revert

File: pmd-apex/src/main/resources/rulesets/apex/style.xml [hunk 9]
```diff
@@ -132,4 +132,24 @@ Many interfaces (e.g. Batch) required global modifiers in the past but don't req
...
+          message="Use of non existent annotations will lead to broken Apex code which will not compile in the future."
```
Comment (jsotuyod) [inline pmd-apex/src/main/resources/rulesets/apex/style.xml]: consider adding the link to the list of valid annotations here for user reference

**Comment by capeterson** [comment]
@jsotuyod I'm the product manager for Apex at Salesforce (Anand is on the dev team), and wanted to answer 

> are we certain it includes all accepted annotations?

This will be the case for the foreseeable future. I can't make an official guarantee as this isn't a formally supported API, but we have no current plans to make changes that would stop this from being true.

File: pmd-apex/src/main/resources/rulesets/apex/style.xml (hunk 9 above)
Comment (jsotuyod) [reply pmd-apex/src/main/resources/rulesets/apex/style.xml]: On second thought, the link should be part of the description rather than the message.

File: pmd-apex/src/main/resources/rulesets/apex/style.xml [hunk 10]
```diff
@@ -132,4 +132,24 @@ Many interfaces (e.g. Batch) required global modifiers in the past but don't req
...
+          class="net.sourceforge.pmd.lang.apex.rule.style.AvoidNonExistentAnnotationsRule"
+          externalInfoUrl="${pmd.website.baseurl}/rules/apex/style.html#AvoidNonExistentAnnotations">
+        <description>
+            Apex supported non existent annotations for legacy reasons.
+            In the future, use of such non-existent annotations could result in broken apex code that will not copile.
+            This will prevent users of garbage annotations from being able to use legitimate annotations added to apex in the future.
```
Comment (jsotuyod) [inline pmd-apex/src/main/resources/rulesets/apex/style.xml]: consider adding the link to the list of valid annotations here for user reference

**Comment by jsotuyod** [comment]
@capeterson thanks for the input! I'm really glad to know that.

I'm looking forward to strengthen the collaboration between PMD and Salesforce. Feel free to contact me to better align objectives / priorities and improve synergies.

**Comment by up2go-rsoesemann** [comment]
@jsotuyod and @capeterson and please always feel free to pull me in as a middle man in case of timing issues. I still feel very responsible ;-)

**Comment by rsoesemann** [comment]
@jsotuyod can I help to solve this conflict? Is there even still a conflict?

**Comment by jsotuyod** [comment]
@rsoesemann there are 2 merge-conflicts and still all my unresolved comments. I don't think you can solve this unless you take the branch into your PMD fork, and submit a new PR yourself.

PMD maintainers *can* do something similar on local copies before merging.

**Comment by rsoesemann** [comment]
@anand13s any reason for not responding to the comments and merge conflicts?

**Comment by capeterson** [comment]
@rsoesemann he's been busy with his day job and all the work I heaped on him for our Summer release. We know we need to fix these before this will be accepted.

**Comment by jsotuyod** [comment]
I personally fixed the remaining issues and merged it. This will be part of PMD 6.5.0.

**Comment by rsoesemann** [comment]
Thanks everybody involved especially @jsotuyod for getting this into production.
